lib/TPP/Dialect/Tpp/TppUtils.cpp

```diff
@@ -138,106 +135,105 @@ bool hasCopySemantics(linalg::LinalgOp linalgOp) {
-// to an another op user currentUser.
-// If there are no earlier users or the specified currentUser is invalid
-// e.g., it does not belong the def-use chain of op, return nullptr instead.
-static Operation *getPrevUser(Operation *op, Operation *currentUser) {
-  if (!op || !currentUser)
-    return nullptr;
-
-  // Iterate over the op users to find the given current user.
-  auto userIt = op->user_begin();
-  while (userIt != op->user_end()) {
-    if (*userIt == currentUser) {
-      // Given that the user_iterator visits the op users from the last to
-      // the first user, simply return the next user after the current user
-      // i.e. an earlier/previous operation.
-      const auto nextUser = ++userIt;
-      return nextUser != op->user_end() ? *nextUser : nullptr;
-    }
+// Returns true if the value is a constant float or integer.
+bool isValConstZero(Value val) {
+  return matchPattern(val, m_AnyZeroFloat()) || matchPattern(val, m_Zero());
+}
+
+// Prototypes
+static bool isZeroOp(Operation *, Operation *);
+
+// Returns true if the value represents a zero filled tensor.
+// Recurse into isZeroOp for defining ops if not imediately obvious
+// Looks past linalg generic's argument (which don't have defining ops)
+static bool isZeroValue(Value val, Operation *userOp) {
+  if (!val)
+    return false;
+  if (isValConstZero(val))
+    return true;
 
-    ++userIt;
+  Operation *defOp = nullptr;
+
+  // Block arguments don't have a defining op, but they do have an op arg
+  if (auto arg = dyn_cast<BlockArgument>(val)) {
+    // We need to find the argument to the linalg on the same order as this one
+    auto *linalgOp = arg.getParentRegion()->getParentOp();
+    if (!isa<linalg::GenericOp>(linalgOp))
+        return false;
+    auto index = arg.getArgNumber();
+    auto linalgArg = linalgOp->getOperand(index);
+    defOp = linalgArg.getDefiningOp();
+  } else {
+    defOp = val.getDefiningOp();
   }
 
-  return nullptr;
+  return isZeroOp(defOp, userOp);
 }
 
-bool isValConstZero(Value val) {
-  return matchPattern(val, m_AnyZeroFloat()) || matchPattern(val, m_Zero());
-}
+// This is a wrapper to isZeroValue as the entry point of the recursive pattern
+bool isZeroTensor(Value val) { return isZeroValue(val, nullptr); }
 
-bool isZeroTensor(Value val) { return isZeroTensor(val.getDefiningOp()); }
+// Returns true if the attribute represent "all zeros"
+bool isZeroAttr(Attribute attribute) {
+  return TypeSwitch<Attribute, bool>(attribute)
+      .Case<FloatAttr>([](auto attr) {
+          return attr.getValueAsDouble() == 0.0;
+      })
+      .Case<IntegerAttr>([](auto attr) {
+          return attr.getInt() == 0;
+      })
+      .Case<DenseElementsAttr>([](auto attr) {
+        if (!attr.getElementType().isIntOrFloat())
+          return false;
+        if (!attr.isSplat())
+          return false;
+        auto value = attr.template getSplatValue<Attribute>();
+        if (auto floatAttr = dyn_cast<FloatAttr>(value)) {
```

Comment:
nit: integer case could also be added for completeness.

Reply:
Better even, recur back into `isZeroAttr`.